### ISSUE TITLE
Code cells with `#| output: asis` ignore `#| tbl-cap` and `#| label:` 

### ISSUE THREAD

**Issue by TS-CUBED** [quarto-dev/quarto-cli#3770]
### Bug description

I try to generate a LaTeX table in Julia using PrettyTables.jl

This creates the LaTeX code for the `\begin{tabular} ... \end{tabular}` sections. 

1. If I use `#| output: true` that gets inserted in a `verbatim` environment, which prints the LaTeX code in the final document, which is not what I want. It does include the `table` environment and adds a caption and label as expected, though. I can use the `@tbl-label` method to cross-reference in this case.

2. If I use `#| output: asis` the output is inserted as is (as expected). This renders a nice table, however, then I don't get the `table` environment and no caption and label.

3. I use 2, and add the `\begin{table}`, `\caption`, `\label` before the block and the `\end{table}` after. Then I get the correct table, incl. caption, but I cannot use the `@tbl-label` to crossreference, which is annoying, since I want to use HTML export as well. For LaTeX `\ref{tbl-label}` works, but then I'd need to add the "Table" in front of that and it would need more trickery to switch to HTML export.

Is there a way to either make (1) not use the verbatim environment, or (2) honour the float options `#| label` and `#| tbl-cap:`?

### Checklist

- [ ] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [ ] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [ ] Please document the operating system you're running. If on Linux, please provide the specific distribution.

**Comment by cscheid** [comment]
Can you please provide a reproducible example?

**Comment by TS-CUBED** [comment]
The following does illustrate the issue. I thought that in the first case I did get a caption and label properly, but I may be mistaken, or made a mistake when I prepared the example. It does not seem to read the `#| tbl-cap:` and `#| label:` properly.

````
---
title: 'Prettytables Test'

license: "CC BY"

format: 
    pdf: 
        keep-tex: true
shift-heading-level-by: 0 
jupyter: julia-1.8

bibliography: /home/thor/Dropbox/Org/library.bib
citation-location: document
---

# Test LaTeX output for table with and without `#| output: asis`

Without `#| output: asis`, the table environment and caption work, but the LaTeX code that would create the table is also enclosed in `verbatim`. 
Referncing @tbl-prettytable with `@tbl-prettytable` works.

```{julia}
#| label: tbl-prettytable
#| tbl-cap: This is the table caption for `asis`

using PrettyTables

headers = ["name", "value"]

data = [
    "test data"       1
    "more data"       1.2
    "getting bored"  2.3
]

pretty_table(data, backend = Val(:latex), header = headers)
```

With `#|output: asis`, the table code gets inserted properly, but the `#| tbl-cap` and `#| label` options are not used and not `table` environment and no caption are created. 
Referncing @tbl-prettytable-asis with `@tbl-prettytable-asis` doesn't work, but `\ref{tbl-prettytable-asis}` does: \ref{tbl-prettytable-asis}.

```{julia}
#| output: asis 
#| label: tbl-prettytable-asis
#| tbl-cap: This is the table caption for `asis`

using PrettyTables

headers = ["name", "value"]

data = [
    "test data"       1
    "more data"       1.2
    "getting bored"  2.3
]

pretty_table(data, backend = Val(:latex), header = headers)
```

Direct Markdown table can be referenced by @tbl-benchmark.


```{julia}
#| echo: false 
#| label: tbl-benchmark
#| tbl-cap: Benchmark times for single model run for $t = 20\tau$
#| tbl-colwidths: [25, 12,12,12,12]
using Markdown

md"""
| $\downarrow$ **Solver** / **Valve** $\rightarrow$ | **Diode** | **Orifice** | **Shi**  | **Mynard** | **SAS**  |
|:--------------------------------------------------|-----------|-------------|----------|------------|----------|
| **Tsit5** (non-stiff)                             | 0.073     | 0.139       | 0.554    | 5.48       | 10.4     |
| **Vern7** (non-stiff)                             | 0.039     | 0.101       | 0.541    | 5.7        | 2.25     |
| **TRBDF2** (stiff)                                | 0.177     | 0.217       | Unstable | 0.395      | Maxiters |
| **Rodas5P** (stiff)                               | 0.265     | 0.815       | Unstable | 2.87       | Unstable |
"""
```
````

**Comment by cscheid** [comment]
Thank you! This is going to take a while for us to fix, but a proper handling of this situation (third-party code emitting raw code that needs crossrefs) is coming (certainly in 2023, but possibly not in, say, January.)

**Comment by cscheid** [comment]
Duplicate of the issue group that contains #3600, #3601, etc.

**Comment by cscheid** [comment]
Ok, I now have a resolution for this (thanks for the patience!).

## tl;dr

````
---
title: 'Prettytables Test'
format: pdf
jupyter: julia-1.9
keep-ipynb: true
keep-md: true
keep-tex: true
---

Referencing @tbl-prettytable-asis with `@tbl-prettytable` works.

```{julia}
#| output: asis 
#| echo: false

using PrettyTables

headers = ["name", "value"]

data = [
    "test data"       1
    "more data"       1.2
    "getting bored"  2.3
]

println("::: {#tbl-prettytable-asis}\n")

println("```{=latex}")
pretty_table(data, backend = Val(:latex), header = headers)
println("```\n")

println("This is the table caption for `asis`")
println("\n:::\n")
```
````

## Why the other attempts won't work

`pretty_table` emits output by printing to stdout. This is not ideal, because jupyter produces the following .ipynb output:

```
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{rr}\n",
      "  \\hline\n",
      "  \\textbf{name} & \\textbf{value} \\\\\\hline\n",
      "  test data & 1 \\\\\n",
      "  more data & 1.2 \\\\\n",
      "  getting bored & 2.3 \\\\\\hline\n",
      "\\end{tabular}\n"
     ]
    }
   ],
```

In order for Quarto to handle this table correctly, we need to read the Markdown output (as you can tell by the `md()` call you use in your last cell).

`output: asis`, on the other hand, completely takes over the rendering of the cell, and doesn't produce _any_ scaffolding. That's by design, but it does mean that the code cell is responsible for producing the correct output. In 1.4, the most robust way to get crossreferenceable artifacts is to use [the div notation](https://quarto.org/docs/prerelease/1.4/crossref.html#crossreferenceable-elements-take-arbitrary-content). That's what my solution does.

In order for PrettyTables to "just work", the library would have to be changed (or perhaps simply configured differently? I'm not a PrettyTables user) to emit output correctly. That could be one of a few possibilities:

### Emit Markdown

The .ipynb would have a cell's output that looks like this:

````
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "| $\\downarrow$ **Solver** / **Valve** $\\rightarrow$ | **Diode** | **Orifice** |  **Shi** | **Mynard** |  **SAS** |\n",
       "|:------------------------------------------------- | ---------:| -----------:| --------:| ----------:| --------:|\n",
       "| **Tsit5** (non-stiff)                             |     0.073 |       0.139 |    0.554 |       5.48 |     10.4 |\n",
       "| **Vern7** (non-stiff)                             |     0.039 |       0.101 |    0.541 |        5.7 |     2.25 |\n",
       "| **TRBDF2** (stiff)                                |     0.177 |       0.217 | Unstable |      0.395 | Maxiters |\n",
       "| **Rodas5P** (stiff)                               |     0.265 |       0.815 | Unstable |       2.87 | Unstable |\n"
      ]
    }
   ]
````

### Emit latex

The .ipynb would have a cell's output that looks like this:

````
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "| $\\downarrow$ **Solver** / **Valve** $\\rightarrow$ | **Diode** | **Orifice** |  **Shi** | **Mynard** |  **SAS** |\n",
       "|:------------------------------------------------- | ---------:| -----------:| --------:| ----------:| --------:|\n",
       "| **Tsit5** (non-stiff)                             |     0.073 |       0.139 |    0.554 |       5.48 |     10.4 |\n",
       "| **Vern7** (non-stiff)                             |     0.039 |       0.101 |    0.541 |        5.7 |     2.25 |\n",
       "| **TRBDF2** (stiff)                                |     0.177 |       0.217 | Unstable |      0.395 | Maxiters |\n",
       "| **Rodas5P** (stiff)                               |     0.265 |       0.815 | Unstable |       2.87 | Unstable |\n"
      ]
    }
   ]
````
